src/StreamJsonRpc/JsonMessageFormatter.cs

```diff
@@ -518,5 +577,88 @@ internal JsonArrayPool(ArrayPool<T> arrayPool)
+
+            public override bool CanConvert(Type objectType) => MessageFormatterHelper.FindIProgressOfT(objectType) != null;
+
+            public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+            {
+                throw new NotSupportedException();
+            }
+
+            public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
+            {
+                // if the requestId is empty it means the Progress object comes from a response
+                if (this.formatter.requestId == null)
+                {
+                    writer.WriteValue(value);
+                    return;
+                }
+
+                long progressId = Interlocked.Increment(ref this.formatter.nextProgressId);
+                this.formatter.requestProgressMap.Add(this.formatter.requestId, progressId);
+
+                this.formatter.progressMap.Add(progressId, value);
+
+                writer.WriteValue(progressId);
+            }
+        }
+
+        private class JsonProgressServerConverter : JsonConverter
+        {
+            private JsonMessageFormatter formatter = null;
+
+            public JsonProgressServerConverter(JsonMessageFormatter formatter)
+            {
+                this.formatter = formatter ?? throw new ArgumentNullException(nameof(formatter));
+            }
+
+            public override bool CanConvert(Type objectType)
+            {
+                return objectType.IsConstructedGenericType && objectType.GetGenericTypeDefinition().Equals(typeof(IProgress<>));
+            }
+
+            public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+            {
+                if (reader.TokenType == JsonToken.Null)
+                {
+                    return null;
+                }
+
+                JToken token = JToken.Load(reader);
+                Type progressType = typeof(JsonProgress<>).MakeGenericType(objectType.GenericTypeArguments[0]);
+                return Activator.CreateInstance(progressType, new object[] { this.formatter.Rpc, token });
+            }
+
+            public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
+            {
+                throw new NotSupportedException();
+            }
+
+            private class JsonProgress<T> : IProgress<T>
+            {
+                private readonly JsonRpc rpc;
+                private readonly JToken token;
+
+                public JsonProgress(JsonRpc rpc, JToken token)
+                {
+                    this.rpc = rpc ?? throw new ArgumentNullException(nameof(rpc));
+                    this.token = token ?? throw new ArgumentNullException(nameof(token));
+                }
+
+                public void Report(T value)
+                {
+                    this.rpc.NotifyAsync("$/progress", this.token, value).Forget();
```

Comment:
>$/progress [](start = 42, length = 10)

Let's move all instances of this magic string (at least within this file) to a single const field.